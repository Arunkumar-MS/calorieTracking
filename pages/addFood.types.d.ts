export interface FoodEntry {
    _id?: string;
    servingQty: number;
    servingUnit: string;
    servingWeightGrams: string;
    url: string;
    calories: number;
    name: string;
    userId: string;
    addedDate: number;
}