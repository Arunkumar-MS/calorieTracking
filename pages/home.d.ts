type Role = 'admin' | 'user' | null;

export interface User {
    name: string;
    userId: string;
    role: Role;
}

export interface Props {
    user: User;
}