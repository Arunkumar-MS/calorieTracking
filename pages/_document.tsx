import Document, { Html, Head, Main, NextScript, DocumentContext } from 'next/document'

class MyDocument extends Document {
  static async getInitialProps(ctx: DocumentContext) {
    const initialProps = await Document.getInitialProps(ctx)
    return { ...initialProps }
  }

  render() {
    return (
      <Html className='h-full bg-gray-100' lang="en-us">
        <title>
          Caloriese tracker
        </title>
        <meta name="description" content="This is caloriese tracking app. which helps you to keep fit by setting caloriese limte per day."/>
        <Head />
        <body className='h-full'>
          <Main />
          <NextScript />
        </body>
      </Html>
    )
  }
}

export default MyDocument