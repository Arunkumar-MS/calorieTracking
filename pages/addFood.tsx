import type { NextPage } from 'next';
import React from 'react';
import Layout from '@Layout/index';
import { useForm } from "react-hook-form";
import { useRouter } from 'next/router';
import {
    useAppDispatch,
} from '@Store/hooks';
import { updateUser } from '@Reducers/userSlice/userSlice';
import { Props as InitailProps } from '@Pages/home.types';
import authAccess from '@Service/serverAuth';
import axios from '@Service/core/axios';
import { FoodEntry } from '@Pages/addFood.types';
import { updateEntry } from '@Reducers/foodDetailsSlice/foodDetailsSlice';

type Props = NextPage & InitailProps;


const saveFood = (data: FoodEntry) => {
    return axios.post('/tracker/addFood', { ...data });
}


const AddFood = (props: Props) => {
    const {
        register,
        handleSubmit,
        formState,
        watch,
        formState: { errors }
    } = useForm();

    const dispatch = useAppDispatch();

    const { query } = useRouter();

    React.useEffect(() => {
        dispatch(updateUser(props.user))
    }, []);

    const onSubmit = (data: any) => {
        saveFood(data).then((list) => dispatch(updateEntry(list.data)));
    };

    const getFormRegisterPropsForString = (name: string) => {
        return {
            ...register(name, {
                required: true,
                maxLength: 150,
                pattern: /^[A-Za-z]+$/i
            })
        }
    }

    const getFormRegisterPropsForURL = (name: string) => {
        return {
            ...register(name, {
                required: true,
                maxLength: 2045,
                pattern: /(http(s)?:\/\/.)?(www\.)?[-a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,6}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)/g
            })
        }
    }

    const getFormRegisterPropsForNumber = (name: string) => {
        return {
            ...register(name, {
                required: true,
                maxLength: 150,
                pattern: /^[0-9]+$/i
            })
        }
    }


    const renderRow = (name: string, registerProps: any, error: any, exp?: string, defultValue?: any, className?: string) => {
        return (
            <div className={`sm:grid sm:grid-cols-3 sm:gap-4 sm:items-start sm:border-t sm:border-gray-200 pt-2 pb-2 sm:pt-5  sm:pb-5 ${className}`}>
                <label htmlFor="last-name" className="block text-sm font-medium text-gray-700 sm:mt-px sm:pt-2">
                    <span>{name}</span>
                    <span className='text-xs'>{exp}</span>
                </label>
                <div>
                    <div className="mt-1 sm:mt-0 sm:col-span-2">

                        <input
                            {...registerProps}
                            type="text"
                            id="last-name"
                            defaultValue={defultValue}
                            className="max-w-lg block w-full shadow-sm focus:ring-indigo-500 focus:border-indigo-500 sm:max-w-xs sm:text-sm border-gray-300 rounded-md"
                        />
                    </div>
                    {error && <span className="flex items-center font-medium tracking-wide text-red-500 text-xs mt-1 ml-1">
                        {`Please enter valid ${name}!`}
                    </span>}
                </div>
            </div>
        )

    }


    return (
        <Layout>
            <form onSubmit={handleSubmit(onSubmit)}>
                {renderRow('Food name', getFormRegisterPropsForString('name'), errors['name'], '(Pizza)', query.foodname)}
                {renderRow('Consume qty  ', getFormRegisterPropsForNumber('consumeQty'), errors['consumeQty'], '(1)')}
                {renderRow('Consumed Calories', getFormRegisterPropsForNumber('consumedCalories'), errors['consumedCalories'], '(120)')}

                {renderRow('Serving unit ', getFormRegisterPropsForString('servingUnit'), errors['servingUnit'], '(Slice)')}
                {renderRow('Consumed weight grams ', getFormRegisterPropsForNumber('consumedWeightGrams'), errors['consumedWeightGrams'], '(107)')}
                {renderRow('Image Url', getFormRegisterPropsForURL('imageUrl'), errors['imageUrl'])}
                <div className=' w-full flex justify-center'>
                    <button type="submit" className="w-full md:w-5/12 mt-2 bg-blue-500 rounded-lg font-bold text-white text-center px-4 py-3 transition duration-300 ease-in-out hover:bg-blue-600">
                        Add
                    </button>
                </div>

            </form>
        </Layout>

    )
}

export async function getServerSideProps(ctx: any) {
    const props = await authAccess(ctx);
    return props;
}

export default AddFood;
