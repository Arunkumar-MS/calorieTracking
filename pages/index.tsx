import type { NextPage } from 'next';
import dynamic from 'next/dynamic';

import React from 'react';
import Layout from '@Layout/index';
import { CalorieTrackerCard } from 'src/component/calorieTrackerCard';
import axios from "axios";
import cookies from 'cookies'
import {
  useAppDispatch,
} from '@Store/hooks';
import { updateUser } from '@Reducers/userSlice/userSlice';
import { Props as HomePageProps } from '@Pages/home';

const UserFlow = dynamic(() => import('../src/component/userFlow/userFlow'));


type Props = NextPage & HomePageProps;

const Home = (props: Props) => {
  const dispatch = useAppDispatch();

  React.useEffect(() => {
    dispatch(updateUser(props.user))
  }, []);

  const renderAdminFlow = () => {

  }
  const renderUserFlow = () => {
    if (props.user.role === 'user') {
      return <UserFlow />
    }
  }

  return (
    <Layout>
      {renderUserFlow()}
      {renderAdminFlow()}
    </Layout>

  )
}

export async function getServerSideProps(ctx: any) {
  try {
    const ck = new cookies(ctx.req, ctx.res)
    const user = await axios.get('http://localhost:3001/auth/getUser', {
      headers: {
        'Content-Type': 'application/json',
        'Authorization': ck.get('token') || '',
      }
    });
    return {
      props: { user: { ...user.data, userId: user.data['_id'] } },
    }
  } catch (e) {
    const { res } = ctx;
    res.setHeader("location", "/login");
    res.statusCode = 302;
    res.end();
    return {
      props: {},
    }
  }
}

export default Home;
