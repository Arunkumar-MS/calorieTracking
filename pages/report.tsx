import React from 'react';
import { selectUser } from '@Reducers/userSlice/userSlice';
import dynamic from 'next/dynamic';
import withAuth from 'src/authHoc';
import { useSelector } from 'react-redux';

const AdminReport = dynamic(() => import('@Component/adminReport'));
const UserReport = dynamic(() => import('@Component/userReport'));


export const ReportComponent = () => {
    const user = useSelector(selectUser);
    const isAdmin = user?.role === 'admin';
    const [selectedTab, setSelectedTab] = React.useState('user');
    const selectedClass = 'border-indigo-500 text-indigo-600 whitespace-nowrap py-4 px-1 border-b-2 font-medium text-sm';
    const nonSelectedClass = 'border-indigo-500 text-indigo-600", Default: "border-transparent text-gray-500 hover:text-gray-700 hover:border-gray-300';
    if (isAdmin) {
        return (
            <>
                <div className="mb-5" data-test-id="report-page-admin-section">
                    <div className="border-b border-gray-200">
                        <nav className="-mb-px flex space-x-8 items-center" aria-label="Tabs">
                            {/* Current: "border-indigo-500 text-indigo-600", Default: "border-transparent text-gray-500 hover:text-gray-700 hover:border-gray-300" */}
                            <a onClick={() => setSelectedTab('user')} href="#" className={selectedTab === 'user' ? selectedClass : nonSelectedClass}>
                                Youre report
                            </a>
                            <a data-test-id="report-page-admin-report" onClick={() => setSelectedTab('admin')} href="#" className={selectedTab === 'admin' ? selectedClass : nonSelectedClass}>
                                All other report
                            </a>
                        </nav>
                    </div>
                </div>
                <div className='w-full md:w-2/4 m-auto text-gray-500 text-center'>
                    {selectedTab === 'admin' && (
                        <AdminReport />
                    )}
                    {selectedTab === 'user' && (
                        <UserReport />
                    )
                    }
                </div>

            </>
        );
    }

    return (
        <>
            <div className='w-8/12 md:w-2/4 m-auto text-gray-500 text-center'>
                <UserReport />
            </div>
        </>

    )
}

const Report = withAuth(ReportComponent, ['admin', 'user']);

export default Report;
