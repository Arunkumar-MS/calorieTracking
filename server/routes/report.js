const express = require("express")
const ObjectId = require('mongodb').ObjectId;
const router = express.Router();
const { getUser } = require('./auth');
const FoodListModel = require('../schema/foodListSchema');
const authorization = require('../middleware/authorization');
const UserModel = require('../schema/userSchema');
const fromUnixTime = require('date-fns/fromUnixTime');
const startOfDay = require('date-fns/startOfDay');
const getUnixTime = require('date-fns/getUnixTime');
const format = require('date-fns/format');
const groupBy = require('lodash.groupby');


router.get("/getuserReport", authorization(['admin', 'user']), async (req, res, next) => {
    const userId = req.ctx.user['_id'];
    try {
        const lastWeekStartDate = getUnixTime(startOfDay(new Date(Date.now() - 6 * 24 * 60 * 60 * 1000)));
        const lastSevenDaysData = await FoodListModel.find({ userId, addedDate: { $gte: lastWeekStartDate } });
        const groupedEntry = lastSevenDaysData.reduce((hashMap, item) => {
            const key = format(fromUnixTime(item.addedDate), 'dd_MM_YYY');
            const value = hashMap[key] || 0;

            hashMap[key] = value + Number(item.consumedCalories);
            return hashMap;
        }, {});
        res.status(200).send(groupedEntry);
        return;
    } catch (e) {
        logger.error('getuserReport: Something went wrong', e);
        res.status(404).send('Bad request!');
        return;
    }
});


router.get("/getAdminReport", authorization(['admin']), async (req, res, next) => {
    const lastWeekStartDate = getUnixTime(startOfDay(new Date(Date.now() - 6 * 24 * 60 * 60 * 1000)));
    const prevWeekStartDate = getUnixTime(startOfDay(new Date(Date.now() - 13 * 24 * 60 * 60 * 1000)));

    const userId = req.ctx.user['_id'];
    try {


        const lastSevenDaysData = await FoodListModel.find({ userId, addedDate: { $gte: lastWeekStartDate } });
        const groupedByDate = lastSevenDaysData.reduce((hashMap, item) => {
            const key = format(fromUnixTime(item.addedDate), 'dd_MM_YYY');
            const value = hashMap[key] || [];
            hashMap[key] = [...value, item];
            return hashMap;
        }, {});

        Object.keys(groupedByDate).map((key) => {

            const dayItems = groupedByDate[key];
            const groupedByUser = groupBy(dayItems, ['userId']);

          Object.keys(groupedByUser).reduce((result, userKey)=> {

            const singleUserEntry = groupedByUser[userKey];
            const totalCal = singleUserEntry.reduce((re,item)=>{
                return re + item.consumedCalories;

            },0);

           


          },0);




        });

        // const groupedEntry = lastSevenDaysData.reduce((groupBy, item) => {
        //     const groupBydateKey = format(fromUnixTime(item.addedDate), 'dd_MM_YYY');
        //     const value = groupBy.groupBydate[groupBydateKey] || 0;
        //     groupBy.groupBydate[groupBydateKey] = value + Number(item.consumedCalories);
        //     const groupByuserKey = item.userId;
        //     const preData = groupBy.groupByuser[groupByuserKey] || [];
        //     groupBy.groupByuser[groupByuserKey] = [...preData, item];
        //     return groupBy;
        // }, { groupBydate: {}, groupByuser: {} });


        const prevLastSevenDaysData = await FoodListModel.find({ userId, addedDate: { $gte: prevWeekStartDate, $lt: lastWeekStartDate } });
        // const groupByuser 
        const result = {
            lastSevenDaysEntries: lastSevenDaysData.length,
            weekBeforeLastSevenDaysEntries: prevLastSevenDaysData.length,
        }
        res.status(200).send(groupedEntry);
        return;
    } catch (e) {
        logger.error('getfoodEntry: Something went wrong', e);
        res.status(404).send('Bad request!');
        return;
    }
});


module.exports = router