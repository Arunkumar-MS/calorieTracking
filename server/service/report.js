const FoodListModel = require('../schema/foodListSchema');
const fromUnixTime = require('date-fns/fromUnixTime');
const startOfDay = require('date-fns/startOfDay');
const getUnixTime = require('date-fns/getUnixTime');
const format = require('date-fns/format');
const groupBy = require('lodash.groupby');

const getuserReport = async (userId, days = 7) => {
    const day = days - 1;
    const lastWeekStartDate = getUnixTime(startOfDay(new Date(Date.now() - day * 24 * 60 * 60 * 1000)));
    const lastSevenDaysData = await FoodListModel.find({ userId, addedDate: { $gte: lastWeekStartDate } });
    const groupedEntry = lastSevenDaysData.reduce((hashMap, item) => {
        const key = format(fromUnixTime(item.addedDate), 'dd:MM:YYY');
        const value = hashMap[key] || 0;

        hashMap[key] = value + Number(item.consumedCalories);
        return hashMap;
    }, {});
    return groupedEntry;
}


const getAdminReport = async (days = 7) => {
    const tillStartOfLastweekday = days - 1;
    const tillPrevLastWeekDate = (days * 2) - 1;
    
    const lastWeekStartDate = getUnixTime(startOfDay(new Date(Date.now() - tillStartOfLastweekday * 24 * 60 * 60 * 1000)));
    const prevWeekStartDate = getUnixTime(startOfDay(new Date(Date.now() - tillPrevLastWeekDate * 24 * 60 * 60 * 1000)));
    const lastSevenDaysData = await FoodListModel.find({ addedDate: { $gte: lastWeekStartDate } });
    const groupedByDate = lastSevenDaysData.reduce((hashMap, item) => {
        const key = format(fromUnixTime(item.addedDate), 'dd:MM:YYY');
        const value = hashMap[key] || [];
        hashMap[key] = [...value, item];
        return hashMap;
    }, {});

    const avgCalForLastSevendays = Object.keys(groupedByDate).map((dateKey) => {
        const dayItems = groupedByDate[dateKey];
        const groupedByUser = groupBy(dayItems, ['userId']);
        const allUsersCalList = Object.keys(groupedByUser).map(userKey => {
            const singleUserEntry = groupedByUser[userKey];
            const totalCal = singleUserEntry.reduce((re, item) => {
                return re + Number(item.consumedCalories);
            }, 0);
            return totalCal;
        }, 0);
        const avgCaloriesAddPerDay = allUsersCalList.reduce((cal, item) => cal + item, 0) / allUsersCalList.length;
        return {
            date: dateKey,
            value: avgCaloriesAddPerDay
        };
    });

    const prevLastSevenDaysData = await FoodListModel.find({ addedDate: { $gte: prevWeekStartDate, $lt: lastWeekStartDate } });

    const result = {
        avgCalForLastSevendays,
        lastSevenDaysEntries: lastSevenDaysData.length,
        weekBeforeLastSevenDaysEntries: prevLastSevenDaysData.length,
    }

    return result;
}

module.exports = {
    getuserReport,
    getAdminReport,
}