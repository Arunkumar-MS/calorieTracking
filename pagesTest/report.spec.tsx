import { jest } from '@jest/globals'
import { shallowToJson } from "enzyme-to-json";
import * as React from 'react';
import { shallow } from 'enzyme';
import { ReportComponent } from '../pages/report';
import * as redux from 'react-redux'
jest
  .spyOn(redux, 'useSelector')
  .mockReturnValueOnce({role: 'user'})
  .mockReturnValueOnce({role: 'admin'})

describe('Report page', () => {
    it('should render properly for normal user', function () {
        jest.clearAllMocks();
        const wrap = shallow(<ReportComponent />)
        expect(shallowToJson(wrap)).toMatchSnapshot();
    })

    it('should render properly for Admin', function () {
        jest.clearAllMocks();
        const wrap = shallow(<ReportComponent />)
        expect(shallowToJson(wrap)).toMatchSnapshot();
    })
})
