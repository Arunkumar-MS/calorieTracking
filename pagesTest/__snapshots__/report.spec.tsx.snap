// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Report page should render properly for Admin 1`] = `
<Fragment>
  <div
    className="mb-5"
    data-test-id="report-page-admin-section"
  >
    <div
      className="border-b border-gray-200"
    >
      <nav
        aria-label="Tabs"
        className="-mb-px flex space-x-8 items-center"
      >
        <a
          className="border-indigo-500 text-indigo-600 whitespace-nowrap py-4 px-1 border-b-2 font-medium text-sm"
          href="#"
          onClick={[Function]}
        >
          Your report
        </a>
        <a
          className="border-indigo-500 text-indigo-600\\", Default: \\"border-transparent text-gray-500 hover:text-gray-700 hover:border-gray-300"
          data-test-id="report-page-admin-report"
          href="#"
          onClick={[Function]}
        >
          All other reports
        </a>
      </nav>
    </div>
  </div>
  <div
    className="w-full md:w-2/4 m-auto text-gray-500 text-center"
  >
    <ForwardRef(LoadableComponent) />
  </div>
</Fragment>
`;

exports[`Report page should render properly for normal user 1`] = `
<Fragment>
  <div
    className="w-8/12 md:w-2/4 m-auto text-gray-500 text-center"
  >
    <ForwardRef(LoadableComponent) />
  </div>
</Fragment>
`;
